### ISSUE TITLE
💬 RFC: Break Sliding Version Policy for the React 18 Upgrade

### ISSUE THREAD

**Issue by RobotSail** [backstage/backstage#18625]
### 🔖 Need

Maintaining the sliding version policy of supporting both React 17 and 18 will cause more headaches for maintainers, consumers, and developers than it will solve. Since many libraries have broken support from 17 to 18, Backstage will have to introduce extra complexity to maintain compatibility between the two versions. This means contributors will need to write more code to work around this inflexibility, and maintainers will have to put in more effort to properly review submissions. 

Additionally, plugin developers will also have to introduce similar logic to maintain compatibility between the two versions, since there may still be people relying on React 17. Besides the code complexity, we have seen [contributors run into issues](https://github.com/backstage/backstage/pull/18322) with Material UI typing due to having support for both React 16 and 17. Finally, developers will have to adjust their codebases to support React 18 anyway, so breaking the versioning policy in this case will make that work simpler for them. 

There are going to be a few places where support for React 17 and 18 will be difficult from a library imports standpoint. To begin with, the way we render components in React 18 will look like:

```tsx
/* rendering the root in React 17 */
import React from 'react';
import ReactDOM from 'react-dom';

const container = document.getElementById('root');
ReactDOM.render(<App />, container);
```

But in React 18, we make the following changes: 

```tsx
import { createRoot } from 'react-dom/client';
const container = document.getElementById('app');
const root = createRoot(container!); 
root.render(<App tab="home" />);
```

The two versions can then be rendered by version checking and dynamically importing the relevant modules:

```tsx
import '@backstage/cli/asset-types';
import React, { type ReactNode } from 'react';
import App from './App';

// render conditionally depending on the react version
const version = React.version.split('.');
const container = document.getElementById('root');
let renderMethod: (element: ReactNode) => void;

// react 18
if (version[0] === '18') {
  const { createRoot } = require('react-dom/client');
  const root = createRoot(container!);
  renderMethod = (el: ReactNode) => root.render(el);
  // react 17
} else {
  const { render } = require('react-dom');
  renderMethod = (el: ReactNode) => render(el, container);
}

renderMethod(<App />);
```

While this doesn’t look appealing on the eyes, it’s still not a very common method to write, since it only needs to exist in a single place realistically. 

Unfortunately, the problem doesn’t just end here. `@testing-library/react-hooks`, an important module which Backstage relies on for many test suites, drops support for React 18 entirely. Instead, the maintainers suggest using `@testing-library/react`. In addition to moving libraries, their testing APIs have also changed, as well as their signatures and how they are expected to be used. 

For example, the old version of `renderHook` from `@testing-library/react-hooks` would return a number of values, such as the error (if any), helper functions to wait until a specific event such as an update or value change, and a waitFor function. The new version of `renderHook` only returns the result of rendering the hook, and `rerender` and `unmount` functions. In order to utilize the other functions, they must be imported from the library:

```diff
- const { result, waitForValueToChange } = renderHook(() =>
+ const { result } = renderHook(() =>
   useEntityRelationNodesAndEdges({
     rootEntityRefs: ['b:d/c'],
     unidirectional: true,
     mergeRelations: true,
   }),
 );
```

As if this wasn’t already inconvenient, the `waitForValueToChange` function no longer exists in the new `@testing-library/react` library whatsoever. The developers now recommend you to instead use the `waitFor` function. Following the above test, the way it would use `waitForValueToChange` would then need to change from:

```tsx
// old method uses this function here
await waitForValueToChange(
    () => result.current.nodes && result.current.edges,
);

// more test logic down here
const { nodes, edges, loading, error } = result.current;

expect(loading).toBe(false);
expect(error).toBeUndefined();
// more expects down here
```
To now instead doing:
```tsx
// now we must keep re-running our expects inside of this function
// until it stops throwing an error
await waitFor(() => {  
	const { nodes, edges, loading, error } = result.current;

	// these will throw
	expect(loading).toBe(false);
	expect(error).toBeUndefined();
	// more expects down here
});
```

Besides this, there is another problem. I mentioned earlier that `renderHooks` no longer returns error as a field. With the new library, you have to explicitly wrap the hook into a function which calls renderHook, and then you call it and assert that it throws an error. Here is an example of how it’s done in React 17 versus 18:

```tsx
// old method:
const { result } = renderHook(
  props =>
    useElementFilter(props.tree, elements =>
      elements
        .selectByComponentData({
          key: WRAPPING_COMPONENT_KEY,
          withStrictError: 'Could not find component',
        })
        .findComponentData({ key: INNER_COMPONENT_KEY }),
    ),
  {
    initialProps: { tree },
    wrapper: Wrapper,
  },
);
expect(result.error?.message).toEqual('Could not find component');
```

In the new method:

```tsx
const run = () =>
  renderHook(
    props =>
      useElementFilter(props.tree, elements =>
        elements
          .selectByComponentData({
            key: WRAPPING_COMPONENT_KEY,
            withStrictError: 'Could not find component',
          })
          .findComponentData({ key: INNER_COMPONENT_KEY }),
      ),
    {
      initialProps: { tree },
      wrapper: Wrapper,
    },
  );

expect(run()).toThrow(new Error('Could not find component'));
```

In theory, supporting all of this functionality while maintaining the version window is possible. Though you can probably sense from the above examples, it would add a lot of additional complexity. This would be a problem for both existing and new contributors alike. Especially when you consider that Backstage is meant to be used by large enterprises, many of which may be coming from non-TypeScript backgrounds such as Java and C#. 

Yet another problem that we uncovered switching dependencies was that certain components would re-render twice with the newer version of React. For example, the following test verifies that components will receive APIs and does so by throwing an error via the errorApi:

```tsx
  it('should provide mock API implementations', async () => {
    const A = () => {
      const errorApi = useApi(errorApiRef);
      errorApi.post(new Error('NOPE'));
      return null;
    };

    const { error } = await withLogCollector(['error'], async () => {
      await expect(renderInTestApp(A)).rejects.toThrow('NOPE');
    });

    expect(error).toEqual([
      expect.objectContaining({
        detail: new Error(
          'MockErrorApi received unexpected error, Error: NOPE',
        ),
      }),
      expect.stringMatching(/^The above error occurred in the <A> component:/),
    ]);
  });
```

In React 18, we encountered a problem where this component would re-render twice, so we would need to check for that error to have been thrown twice. After investigating, we found that the only way to avert this behavior without rewriting any of the code was by forcing the testing library to render with synchronous mode by passing `legacyRoot: true` to the `render` method from `@testing-library/react`. This causes the component to fallback to using the rendering method for React 17. Although this is a fixable problem, it illuminates another point where messy logic would have to be introduced in order to compensate for the fact that we are supporting multiple React versions. Having to fallback to legacy rendering in tests would also mean missing behavior that would exist when someone is running Backstage with React 18 – a significant portion of the user base.

Another issue you’ll encounter maintaining compatibility is that, because `@testing-library/react` drops support for React 17 altogether, we would need to take some extra steps in order for it to pull a React-17 compatible version. Because this is a major breaking change from their side, the APIs used from that package would change depending on the respective version of React being used. For reference, this is the number of tests affected by this change:

![image](https://github.com/backstage/backstage/assets/97077423/cdac3f8f-8847-48a5-b597-a96927b34e91)
(yes I am using a light theme in this screenshot and proud of it)

### 🎉 Proposal

We propose the following:

1. We will break the version window policy in order to avert adding additional complexity to maintain a legacy React version. While this isn’t ideal, it will help prevent headaches for many contributors and maintainers. Since we will only be worrying about a single version of React, debugging React-specific issues will be straightforward, because we’ll be able to rule out any legacy behavior. This will ease the burden for new adopters as well, because they will not have to write their components and tests with multiple versions in mind. Maintainers will also benefit from this, since the code will be simpler to read, reviewing PRs will be much easier. Since many libraries already require React 18 anyway such as Material UI v5, much of the code will need to be migrated to use React 18 regardless. Breaking the version window will simply reduce the amount of code that will need to be written for it.
1. We will do our best to provide resources in order to assist anyone still on React 17 migrate their codebase and test suites to the new version. We can contribute any tooling that will be necessary here, provide documentation on what to upgrade, how to upgrade, and assist community members with the upgrade.

### 〽️ Alternatives

We have considered maintaining compatibility with both React 17 and 18, despite the complexity and the overhead it would introduce. The alternative would involve:

1. **Conditional Rendering**: By writing conditional rendering logic, we could support both React 17 and 18. However, this could make our code harder to maintain and less readable. Moreover, the cases where the logic is required are rare, which makes the effort seem disproportionate.
1. **Testing Libraries**: We considered keeping the `@testing-library/react-hooks` for React 17 and using `@testing-library/react` for React 18. However, this alternative would force us to adapt to two different testing APIs, and it might also imply duplicate tests or conditional testing logic.
1. **Legacy Rendering**: Using `legacyRoot: true` in tests to fall back to React 17 rendering could avoid the double rendering issue in React 18. However, this would also mean our tests wouldn't accurately represent the behavior seen by a significant portion of our user base running Backstage with React 18.  <ins>**Note**</ins>: This will only help with tests, not the main app, so it’s not a complete alternative.  

### ❌ Risks

The primary risks associated with this proposal include:

1. **Community Acceptance**: There's a risk of causing friction within the community, particularly for those who still rely on React 17. These users would be compelled to upgrade to React 18.

2. **Testing Coverage**: By moving away from legacy rendering in tests, there is a potential risk of uncovering new, unanticipated issues that could affect the stability of our codebase.

3. **Migration Complexity**: The migration process from React 17 to 18 might be challenging for some users, particularly those new to Backstage or from non-TypeScript backgrounds.

4. **Dependency Problems**: There's also a risk associated with the dependencies, especially for frontend plugins, that have broken compatibility between React 17 and 18, which might introduce unknown challenges that we have yet to anticipate.

5. **Future Precedent**: This move could set a precedent for breaking the sliding version window policy for future upgrades, which could lead to uncertainty about our versioning policy.

It feels like it’s possible to have both React 17 & 18 code in the same codebase. The problem is that I can sense it introducing lots of complexity and giving maintainers a death by a thousand cuts. I also feel like for many dependencies, this isn’t a problem, because with something like an Apollo GraphQL client, it only gets used so much. But since React is a fundamental part of the codebase, it becomes a much more intensive thing to maintain. 

I’m sure that there’s a lot of stuff that can be tested to see what works / what doesn’t, but the problem is that it has to be done in the first place.

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by awanlin** [comment]
We upgraded to React 18 in late May of this year. It was a pretty smooth migration and has been smooth sailing ever since. We would consider our Backstage instance about mid-sized with little over a dozen custom frontend plugins just to give you a sense of size. Where we have a gap is our test coverage isn't as good as we would like and the most work for this migration was around fixing tests.

**Comment by Rugvip** [comment]
Summary of discussion that we had last week around this: We're still hopeful that we'll be able to ship support for v18 without breaking compatibility with v17. We would still do an immediate switch to v18 in this repo and stop testing towards v17, but that doesn't mean we need to break v17 support. In the few places where it might be required, such as test utilities, we will provide best effort multi-version support without tests. The support for v17 may slowly or immediately break over time after this initial release, but we will aim to have at least a few releases that provide a window for projects to migrate to v18 at their own pace.

**Comment by camilaibs** [comment]
Thanks, everyone for the comments! This seems to be addressed by #20598, right?
We're closing this issue, but feel free to reopen it if that's not the case 🙌🏻
